Kube-proxy/ipvs; Use go "net" lib to get nodeIPs

uablrek opened the pull request:
#### What type of PR is this?

/kind bug
/kind cleanup

#### What this PR does / why we need it:

To get node addresses to be used for nodePorts using netlink to read the local routing table was unnecessary complex and caused problems with older kernels as reported in #93858.

Using the go `net` library makes the code cleaner and more maintainable while reducing dependencies.

#### Which issue(s) this PR fixes:

Fixes #93858

#### Special notes for your reviewer:

I have limit the update to one file.

The "NetLinkHandle" in the `realIPGetter` is not really needed but unfortunately the unit tests are using the `realIPGetter` with a fake NetLinkHandle. So to limit the size of the PR (and the risk of mistakes) I let it be.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

N/A

uablrek commented:
/retest

aojea commented on pkg/proxy/ipvs/netlink_linux.go:
```diff
@@ -123,72 +124,67 @@ func (h *netlinkHandle) ListBindAddress(devName string) ([]string, error) {
+	return ips, nil
+}
+
+func (h *netlinkHandle) appendAddresses(ips sets.String, ifi *net.Interface) {
```
you don't need this to be a method, actually none of those, but I assume you want them to keep compatible the code

uablrek replied:
I actually wrote all logic into `proxier.go` but then I realized that would force me to re-write the unit-tests quite extensively. I had to balance the work a bit.

aojea commented on pkg/proxy/ipvs/netlink_linux.go:
```diff
@@ -123,72 +124,67 @@ func (h *netlinkHandle) ListBindAddress(devName string) ([]string, error) {
-	for _, route := range routes {
-		if route.LinkIndex == filterLinkIndex {
+	ips := sets.NewString()
+	for _, i := range ifaces {
```
I always struggle with this , but I think that you have to capture this variable since you are passing it as reference
https://github.com/golang/go/wiki/CommonMistakes#using-reference-to-loop-iterator-variable
```
i := i
```

uablrek replied:
I think it's ok in this case. The interface pointer &i is only used within the loop to get the addresses. It's not stored anywhere.

aojea replied:
yeah, but `h.appendAddresses(ips, &i)`  will only use the last value, check this example from the link I pasted
```
While this example might look a bit obvious, the same unexpected behavior could be more hidden in some other cases. For example, the loop variable can be an array and the reference can be a slice:
func main() {
	var out [][]int
	for _, i := range [][1]int{{1}, {2}, {3}} {
		out = append(out, i[:])
	}
	fmt.Println("Values:", out)
}
Output:

Values: [[3] [3] [3]]
```

uablrek replied:
But in that example you have actually stored an address in `out`. In `h.appendAddresses(ips, &i)` the pointer is just used once.

uablrek replied:
And I made some printouts when testing and addresses from all interfaces are checked and added.

aojea replied:
I told you that I always struggle with it XD, if you've tested it has to be ok

aojea commented on pkg/proxy/ipvs/netlink_linux.go:
```diff
@@ -123,72 +124,67 @@ func (h *netlinkHandle) ListBindAddress(devName string) ([]string, error) {
+	}
+	return ips, nil
+}
+
```
a personal thought, is not clearer to have a function
```
func getInterfaceAddresses(ifi *net.Interface) []string {
}
```
and do the set insertions in the main loop?

uablrek replied:
Good point. I'll check that :+1:

uablrek replied:
Would remove the reference-issue also.

aojea replied:
checking in, do you prefer current format and no try to return a slice instead of  set?

aojea commented on pkg/proxy/ipvs/netlink_linux.go:
```diff
@@ -123,72 +124,67 @@ func (h *netlinkHandle) ListBindAddress(devName string) ([]string, error) {
 }
 
-// GetLocalAddresses lists all LOCAL type IP addresses from host based on filter device.
-// If dev is not specified, it's equivalent to exec:
-// $ ip route show table local type local proto kernel
-// 10.0.0.1 dev kube-ipvs0  scope host  src 10.0.0.1
-// 10.0.0.10 dev kube-ipvs0  scope host  src 10.0.0.10
-// 10.0.0.252 dev kube-ipvs0  scope host  src 10.0.0.252
-// 100.106.89.164 dev eth0  scope host  src 100.106.89.164
-// 127.0.0.0/8 dev lo  scope host  src 127.0.0.1
-// 127.0.0.1 dev lo  scope host  src 127.0.0.1
-// 172.17.0.1 dev docker0  scope host  src 172.17.0.1
-// 192.168.122.1 dev virbr0  scope host  src 192.168.122.1
-// Then cut the unique src IP fields,
-// --> result set: [10.0.0.1, 10.0.0.10, 10.0.0.252, 100.106.89.164, 127.0.0.1, 172.17.0.1, 192.168.122.1]
-
-// If dev is specified, it's equivalent to exec:
-// $ ip route show table local type local proto kernel dev kube-ipvs0
-// 10.0.0.1  scope host  src 10.0.0.1
-// 10.0.0.10  scope host  src 10.0.0.10
-// Then cut the unique src IP fields,
-// --> result set: [10.0.0.1, 10.0.0.10]
-
-// If filterDev is specified, the result will discard route of specified device and cut src from other routes.
```
I found these examples useful

uablrek replied:
I'll bring them back. But there is a copy in `proxier.go` I think.

uablrek replied:
The comments are back

uablrek commented:
/retest

uablrek commented:
/retest

aojea commented on pkg/proxy/ipvs/netlink_linux.go:
```diff
@@ -143,52 +148,62 @@ func (h *netlinkHandle) ListBindAddress(devName string) ([]string, error) {
 		if h.isIPv6 {
-			if route.Dst.IP.To4() == nil && !route.Dst.IP.IsLinkLocalUnicast() {
-				res.Insert(route.Dst.IP.String())
+			if ip.To4() != nil || ip.IsLinkLocalUnicast() {
```
this applies to ipv4 to, do you want to discard them only for ipv6?
https://golang.org/src/net/ip.go?s=4256:4294#L143

uablrek replied:
I keep the link-local check for ipv6 only. The thing is that for ipv6 a "scope" is required for link-local addresses, which is an interface, like `fe80::200:ff:fe01:1%eth0` and they are usually auto-assigned so they are always there. For ipv4 it would be possible to use them I think, and whey must have been configured by some one.

aojea commented on pkg/proxy/ipvs/netlink_linux.go:
```diff
@@ -143,52 +148,62 @@ func (h *netlinkHandle) ListBindAddress(devName string) ([]string, error) {
 		if h.isIPv6 {
-			if route.Dst.IP.To4() == nil && !route.Dst.IP.IsLinkLocalUnicast() {
-				res.Insert(route.Dst.IP.String())
+			if ip.To4() != nil || ip.IsLinkLocalUnicast() {
```
you can simplify the check with the utitlnet.IsIPv6 from net/utils and avoid nested ifs
```
if h.isIPv6 != utilnet.IsIPv6(ip) {
   continue
}
if h.isIPv6 && ip.IsLinkLocalUnicast() {
   continue
}
```

uablrek replied:
Better! Thanks. I will update

aojea commented:
LGTM
/assign @andrewsykim 
for final approval

uablrek commented:
/retest

aojea commented:
/hold

I need to apologise, seems that using the go net library has performance issues
https://github.com/kubernetes/kubernetes/pull/103116

uablrek commented:
Re-factored with `net.InterfaceAddrs()`.

uablrek commented:
/test pull-kubernetes-e2e-gci-gce-ipvs

aojea commented:
/lgtm

This was stuck for a long time, if after some days nobody else review I think we can just unhold it

andrewsykim commented on pkg/proxy/ipvs/netlink_linux.go:
```diff
@@ -143,52 +148,63 @@ func (h *netlinkHandle) ListBindAddress(devName string) ([]string, error) {
 // --> result set: [10.0.0.1, 10.0.0.10]
-
-// If filterDev is specified, the result will discard route of specified device and cut src from other routes.
 func (h *netlinkHandle) GetLocalAddresses(dev, filterDev string) (sets.String, error) {
```
Given this is a big refactor for a (very) important function, any thoughts on adding some unit tests for it?

uablrek replied:
Unit test existed but IMHO they mostly test the fake implementation. With the new implementation I have added unit-tests for the `IpIsValidForSet` function.

andrewsykim commented on pkg/proxy/ipvs/netlink_linux.go:
```diff
@@ -143,52 +148,63 @@ func (h *netlinkHandle) ListBindAddress(devName string) ([]string, error) {
+		return h.getLocalAddressesDev(dev)
+	}
+
+	if filterDev == "" {
```
It seems like empty filterDev was valid but now we're erroring here. Do you know if we never pass in empty filterDev here?

aojea replied:
AFAIk this is only used in https://github.com/kubernetes/kubernetes/blob/e6922078577c5c8896b3619857dd744f1c1ecf6b/pkg/proxy/ipvs/proxier.go#L305-L324

uablrek replied:
I have now made the _really big_ refactoring that I wanted to do from the start. I do not like "combo" functions with clever use of parameters to make the function do different things. @andrewsykim IMHO your question proves my point.

@aojea Sorry, I have removed the big address comment again. It is no longer relevant.

uablrek commented:
I have added the function commented in https://github.com/kubernetes/kubernetes/pull/103116#issuecomment-870334408 with unit-tests. This makes further code clean-ups possible. The address handling code is very messy with conversions back ans forth from strings to net.IP and from slices and sets.

uablrek commented:
/test pull-kubernetes-e2e-gci-gce-ipvs

uablrek commented:
/retest

uablrek commented:
/remove-area apiserver

uablrek commented:
/remove-sig node

uablrek commented:
/remove-sig apps
/remove-sig api-machinery
/remove-sig auth
/remove-sig cli
/remove-sig cloud-provider
/remove-sig cluster-lifecycle
/remove-sig instrumentation
/remove-sig node
/remove-sig release
/remove-sig scalability
/remove-sig scheduling
/remove-sig testing

uablrek commented:
I accidently pushed a merge commit. Sorry for the mess you other sig-folks :anguished:

uablrek commented:
/remove-area code-generation
/remove-area dependency
/remove-area kubeadm
/remove-area kubectl
/remove-area kubelet
/remove-area network-policy
/remove-area provider/gcp
/remove-area release-eng
/remove-area test

uablrek commented:
/remove-kind api-change

uablrek commented:
/retest

uablrek commented:
/area kube-proxy

uablrek commented:
@ehashman Please remove the sig/node triage tag. I accidently pushed a merge commit and the robot added just every sig and area code in existence.

uablrek commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

Commit:
Kube-proxy/ipvs; Use go "net" lib to get nodeIPs

The nodeIPs to be used for nodePorts were collected using
netlink which was unnecessary complex and caused se #93858

uablrek commented:
I have to check some filtering functions so hold review for now.

aojea commented:
/test pull-kubernetes-e2e-gci-gce-ipvs

aojea commented:
/hold cancel

aojea commented on pkg/proxy/ipvs/proxier_test.go:
```diff
@@ -34,6 +34,8 @@ import (
 	"k8s.io/apimachinery/pkg/util/intstr"
 	"k8s.io/apimachinery/pkg/util/sets"
 	utilfeature "k8s.io/apiserver/pkg/util/feature"
+
+	//utilproxy "k8s.io/kubernetes/pkg/proxy/util"
```
is this a leftover?

uablrek replied:
Yes. Fixed.

aojea commented on pkg/proxy/util/utils.go:
```diff
@@ -251,6 +251,29 @@ func GetNodeAddresses(cidrs []string, nw NetworkInterfacer) (sets.String, error)
+	IpIsValidForSet(ip net.IP) bool
+}
+
+func AddressSet(validator IpSetValidator, addrs []net.Addr) sets.String {
```
we should add a comment explaining what this function does

uablrek replied:
I re-worked the function. The validator is now a function, which is much better. I did not know that you can pass a method as a function in go (you can't do that in C++).

aojea commented on pkg/proxy/ipvs/netlink_linux.go:
```diff
@@ -123,72 +125,45 @@ func (h *netlinkHandle) ListBindAddress(devName string) ([]string, error) {
-		}
-		filterLinkIndex = link.Attrs().Index
-	}
+type familyValidator struct {
```
we should add the method just next instead at the end of the file
Also, add a statement for the compiler to check it matches the interface

```

Assuming the question is about Go, e.g.
```go
var _ util.IpSetValidator = familyValidator{}
```

uablrek replied:
I re-worked the AddressSet() function, please see comment above.

uablrek commented:
Please note; I re-worked the AddressSet() function. The validator is now a function not an interface.

aojea commented on pkg/proxy/ipvs/netlink_linux.go:
```diff
@@ -124,72 +126,41 @@ func (h *netlinkHandle) ListBindAddress(devName string) ([]string, error) {
-		} else if route.Src != nil {
-			res.Insert(route.Src.String())
-		}
+	if ip.IsLoopback() {
```
should we mention in the comments that loopbacks are already excluded, it only mentions about ipv6 link-locals

aojea replied:
I mean in `pkg/proxy/ipvs/netlink.go`

uablrek replied:
Updated

aojea commented on pkg/proxy/ipvs/proxier_test.go:
```diff
@@ -430,12 +426,12 @@ func TestGetNodeIPs(t *testing.T) {
 		},
 		// case 9
 		{
-			devAddresses: map[string][]string{"ipvs0": {"1.2.3.4"}, "lo": {"127.0.0.1"}, "encap7": {"10.20.30.31"}},
```
why removing the loopbacks here, shouldn't those be excluded by the new methods `GetAllLocalAddresses()`?

uablrek replied:
Because the GetAllLocalAddresses() comes from the "fake" in testing/fake.go, and that does not filter loopback (code was moved). I could add filtering in "fake", but then I would like test the test-code.

uablrek replied:
But you are right. It is clearer to have it in the test-case if the mock procedure would change in the future. I'll restore it and add some ipv6 tests for good measure.

But I will not be able to do that today.

uablrek commented:
@aojea I have problems with running the typecheck scripts locally;
```
kubernetes > ./hack/verify-typecheck-dockerless.sh
type-checking windows/arm64
type-checking linux/arm
type-checking windows/386
signal: killed
!!! Type Check has failed. This may cause cross platform build failures.
!!! Please see https://git.k8s.io/kubernetes/test/typecheck for more information.
```
A re-run can help, but not always. Do you know the cause and some way to fix this?

aojea commented on pkg/proxy/ipvs/netlink_linux.go:
```diff
@@ -21,8 +21,10 @@ package ipvs
+	"net"
 
 	"k8s.io/apimachinery/pkg/util/sets"
+	utilproxy "k8s.io/kubernetes/pkg/proxy/util"
```
this seems the reasons that verify check breaks, since it doesn't compile in windows

> ERROR(windows/arm64): /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/proxy/ipvs/proxier_test.go:5429:9: undeclared name: netlinkHandle
> ERROR(windows/arm64): /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/vishvananda/netlink/xfrm.go:13:36: IPPROTO_ROUTING not declared by package unix
> ERROR(windows/arm64): /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/vishvananda/netlink/xfrm.go:14:36: IPPROTO_ESP not declared by package unix

aojea commented on pkg/proxy/ipvs/proxier_test.go:
```diff
@@ -27,6 +27,7 @@ import (
 	"time"
 
 	"github.com/stretchr/testify/assert"
+	"github.com/vishvananda/netlink"
```
this seems the reasons that verify check breaks, since it doesn't compile in windows

ERROR(windows/arm64): /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/proxy/ipvs/proxier_test.go:5429:9: undeclared name: netlinkHandle
ERROR(windows/arm64): /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/vishvananda/netlink/xfrm.go:13:36: IPPROTO_ROUTING not declared by package unix
ERROR(windows/arm64): /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/vishvananda/netlink/xfrm.go:14:36: IPPROTO_ESP not declared by package unix

uablrek commented:
Squashed and force-pushed (finally!)

aojea commented:
/lgtm
/assign @thockin

thockin commented:
@uablrek did you make this commit with different email maybe?  Not sure why CLA is failing...

thockin commented:
Thanks!

/lgtm
/approve

aojea commented:
> @uablrek did you make this commit with different email maybe? Not sure why CLA is failing...

the sig-contrib? is rolling out the new EasyCLA, there was an announcement somewhere, I remember they mention there can be issues with some users

uablrek commented:
I have commented it on slack https://kubernetes.slack.com/archives/C09R23FHP/p1636757607042000
And updated the feedback issue https://github.com/kubernetes/org/issues/3068#issuecomment-967802772

The merge should work anyway I am told. This seem to be a test that it's working. However there doesn't seem to be any fixing going on when it doesn't work, as for me and some others.
